_posts/2019-04-24-server-rendering-responsively.md

```diff
@@ -0,0 +1,199 @@
+---
+layout: epic
+title: "Server-Rendering A Responsive Artsy.Net"
+date: "2019-04-22"
+author: [steve-hicks, eloy]
+categories: [react, html, web]
+---
+
+At Artsy, we use server-side rendering (SSR) to deliver every page you hit on [artsy.net](https://artsy.net). We
+decided on using SSR for many reasons, amongst them performance. (TODO: Maybe more about why we chose SSR, but I
+don't know much about that history.)
+
+(TODO: Maybe a bit more about SSR & `hydrate`, or just link to those docs.)
+
+One challenge we've run into with SSR is rendering markup and styles from the server that the browser doesn't have
+to throw away. When your browser finishes loading our JavaScript & hydrates our React app, our app suddenly has
+more information about your browser dimensions than we had in the server request. React will go through the
+reconciliation process with this new information, to determine which DOM elements need to be updated. If there are
+significant differences between the markup our server gave you and the component tree our React app is now
+rendering, there's noticeable churn in the DOM.
+
+We'd like our app to churn as little as possible during hydration. We've taken several steps (and mis-steps) toward
+accomplishing this goal: how can we emit _usable_ responsive markup from the server, so that the browser doesn't
+have to throw everything away when the browser dimensions become known?
```

Comment:
With SSR and React 16 it's less about churn as React will hydrate whatever you throw at it, and not patch, as it's more permissive and intelligent than React 15. That said, it's important that one tries to keep the markup rendered on the server identical to what is then eventually rendered on the client lest hydration warnings (and other error's appear).

With React Responsive media the important point is that it permits fine-grained control over rendering exact breakpoints, vs approximating the screen-size on the server (via user-agent detection) and then potentially delivering an incorrect layout. 

RRM, during the server-side pass, will render ALL breakpoint content, and on client-side mount the client then knows exactly what to use, throwing the rest away. So in a sense there's _more_ churn with RRM, as there's a good amount of html that is then disposed of once things settle. But it is is able, with exactitude, to deliver the exact layout without the need for flimsy server-side detection. 

This is an innovative solution to a problem that has been around for a long time! Important though to describe the mechanism right since theres potentially a lot of markup being sent over from the server that will eventually be discarded.

Reply:
Note! Reading further down after the comment above I see how things evolve through iteration. It seems like some kind of summary of conclusions would be appropriate though which is then elaborated upon within each iteration section, connecting the dots.

Reply:
I agree on that approach. I'll work on it!